### DISCUSSION TITLE
How do you configure several registry URLs for NPM datasource?

### DISCUSSION THREAD

**Discussion by taufort** [renovatebot/renovate#31205] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab

### Please tell us more about your question or problem

Hi,

I am using Renovate version 38.23.3.

I'm trying to onboard some NPM projects with Renovate and those projects get their dependencies from several registries. Unfortunately, I discovered recently that NPM datasource does not support having several registry URLs it would seem. Only the first URL in the array registryUrls is caught and used by Renovate. Here is my configuration:

```
  "packageRules": [
    {
      "matchDatasources": ["npm"],
      "registryUrls": [
        "https://mydomain.com/repository/fubar/",
        "https://mydomain2.com/repository/fubar2/",
        "https://mydomain2.com/repository/fubar3/"
      ],
      "addLabels": ["npm"]
    }
  ]
```

In this example, URLs "https://mydomain2.com/repository/fubar2/" and "https://mydomain2.com/repository/fubar3/" are totally ignored by Renovate and Renovate warns me in onboarding MRs that some dependencies could not be resolved. I even saw a WARN log in my self-hosted Renovate pipeline that warns be about this behavior (see logs section).

After rechecking the documentation, I saw that the registryUrls field "supports multiple URLs but it is datasource-dependent on whether only the first is used or multiple." (see [doc](https://docs.renovatebot.com/configuration-options/#registryurls)). I was surprised by this behavior because I've previously set up several registry URLs for maven projects and it worked like a charm. 

Is there a way around this specific behavior with registry URLs for the NPM datasource?

### Logs (if relevant)

<details><summary>Logs</summary>

```
 WARN: Excess registryUrls found for datasource lookup - using first configured only (repository=fubar)
       "datasource": "npm",
       "packageName": "@commitlint/cli",
       "registryUrls": [
         "https://mydomain.com/repository/fubar/",
         "https://mydomain2.com/repository/fubar2",
         "https://mydomain2.com/repository/fubar3"
       ]
```

</details>

**Comment by rarkins** [comment]
In the npm ecosystem there is only one default registry (this is a npm limitation, not Renovate), but you can have alternative registries per _scope_. e.g. `@fubar2/` scope could point to `https://mydomain2.com/repository/fubar2/`. Do you typically use a `.npmrc` file to define how to map these registries?

**Reply by taufort** [reply]
Hi @rarkins and thanks for your reactivity, very much appreciated!

Yes, we use a `.npmrc` file in the NPM projects we want Renovate to onboard and we indeed use the scope mechanic you talked about as such:

```
registry=https://mydomain.com/repository/fubar/
other_dependency=https://mydomain2.com/repository/fubar2/
```

**Reply by rarkins** [reply, marked as answer]
If there is a `.npmrc` in the repo then you shouldn't need to make any configuration change in Renovate config.

Alternatively if repos are missing a `.npmrc` then you could specify `npmrc` in a global preset which is then shared by all, with probably npmrcMerge=true to allow repos to override

**Reply by taufort** [reply]
Thanks a lot for your feedback @rarkins, I had time to test your proposal yesterday and it worked like a charm. Here is my `npm.json5` configuration for those having the same issue I had:

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "npmrc": "registry=https://mydomain.com/repository/fubar/\n@app:registry=https://mydomain2.com/repository/fubar2/\n@app2:registry=https://mydomain2.com/repository/fubar3/\n@app3:registry=https://mydomain2.com/repository/fubar3/",
  "npmrcMerge": "true",
  "packageRules": [
    {
      "matchDatasources": ["npm"],
      "addLabels": ["npm"]
    }
  ]
}

```

**Reply by rarkins** [reply]
Thanks for sharing your solution to benefit others!
